File: Problems/117. Populating Next Right Pointers in Each Node II.py
"""
Problem Link: https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/description
"""

"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if (root == None):
            return

        queue = [root]

        while (queue):
            prev = None
            size = len(queue)
            while (size > 0):
                current = queue.pop(0)
                if (prev != None):
                    prev.next = current
                prev = current
                if (current.left != None):
                    queue.append(current.left)
                if (current.right != None):
                    queue.append(current.right)
                size -= 1
        return root

File: Problems/117. Populating Next Right Pointers in Each Node II.java
/*
Problem Link: https://leetcode.com/problems/populating-next-right-pointers-in-each-node-ii/description
*/

/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

import java.util.*;

class Solution {
    public Node connect(Node root) {
        if (root == null)
            return null;
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            int size = q.size();
            Node prev = null;
            while (size-- > 0) {
                Node curr = q.poll();
                if (prev != null)
                    prev.next = curr;
                prev = curr;
                if (curr.left != null)
                    q.add(curr.left);
                if (curr.right != null)
                    q.add(curr.right);
            }
        }
        return root;
    }
}     
                     
                     
                    